sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,566 @@
+    local_nb_iterations = math.floor(math.log2(work_group_size))
+    f32zero = float32(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        group = dpex.get_group_id(0)
+        thread = dpex.get_local_id(0)
+        first_thread = thread == 0
+
+        n = v.shape[0]
+
+        shm = dpex.local.array(work_group_size, dtype=float32)
+
+        group_start = group * work_group_size * 2
+        thread_operand_1 = group_start + thread
+        thread_operand_2 = group_start + work_group_size + thread
+
+        if thread_operand_1 >= n:
+            shm[thread] = f32zero
+        elif thread_operand_2 >= n:
+            shm[thread] = v[thread_operand_1]
+        else:
+            shm[thread] = v[thread_operand_1] + v[thread_operand_2]
+
+        dpex.barrier()
+        current_size = work_group_size
+        for i in range(local_nb_iterations):
+            current_size = current_size // 2
+            if thread < current_size:
+                shm[thread] += shm[thread + current_size]
+
+            dpex.barrier()
+
+        if first_thread:
+            w[group] = shm[0]
+
+    _steps_data = []
+    n_groups = n
+    while n_groups > 1:
+        n_groups = math.ceil(n_groups / (2 * work_group_size))
+        global_size = n_groups * work_group_size
+        _steps_data.append(
+            (
+                sum_reduction_kernel[global_size, work_group_size],
+                dpctl.tensor.empty(n_groups, dtype=np.float32, device=device),
+            )
+        )
+
+    def sum_reduction(v):
+        for sum_fn, w in _steps_data:
+            sum_fn(v, w)
+            v = w
+        return v
+
+    return sum_reduction
+
+
+@cached_kernel_factory
+def make_sum_reduction_kernel_2(n, dim, work_group_size):
+    global_size = math.ceil(n / work_group_size) * work_group_size
+    f32zero = float32(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        thread = dpex.get_global_id(0)
+        if thread >= n:
+            return
+        tmp = f32zero
+        for d in range(dim):
+            tmp += v[d, thread]
+        w[thread] = tmp
+
+    return sum_reduction_kernel[global_size, work_group_size]
+
+
+@cached_kernel_factory
+def make_sum_reduction_kernel_3(dim0, dim1, dim2, work_group_size):
+    nb_groups_for_dim2 = math.ceil(dim2 / work_group_size)
+    nb_threads_for_dim2 = nb_groups_for_dim2 * work_group_size
+    global_size = nb_threads_for_dim2 * dim1
+    f32zero = float32(0.0)
+
+    @dpex.kernel
+    def sum_reduction_kernel(v, w):
+        group = dpex.get_group_id(0)
+        thread = dpex.get_local_id(0)
+        d1 = group // nb_groups_for_dim2
+        d2 = ((group % nb_groups_for_dim2) * work_group_size) + thread
+        if d2 >= dim2:
+            return
+        tmp = f32zero
+        for d in range(dim0):
+            tmp += v[d, d1, d2]
+        w[d1, d2] = tmp
+
+    return sum_reduction_kernel[global_size, work_group_size]
+
+
+@cached_kernel_factory
+def make_fused_kernel_fixed_window(
```

Comment:
I would always put kernel at the end of the kernel function names.

```suggestion
def make_fused_fixed_window_kernel(
```